lib/Conversion/TorchToLinalg/Random.cpp

```diff
@@ -154,6 +160,104 @@ class ConvertAtenUniformOp : public OpConversionPattern<AtenUniformOp> {
+         pinMemory)) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: pin_memory must be either None or false");
+    }
+
+    SmallVector<int64_t> outSize;
+    if (!matchPattern(op.size(), m_TorchConstantIntList(outSize)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int size list");
+
+    Value initTensor =
+        rewriter.create<tensor::EmptyOp>(loc, outSize, resultElemTy);
+
+    Value multiplier = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getI64IntegerAttr(6364136223846793005));
+    Value incrementStep = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getI64IntegerAttr(1442695040888963407));
+
+    // Get initial seed, min and max used by `linalg.generic` compute payload.
+    Value initialSeed = rewriter.create<TorchConversion::GetNextSeedOp>(loc);
+    Value min = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getFloatAttr(resultElemTy, 0.0));
+    Value max = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getFloatAttr(resultElemTy, 1.0));
+
+    Value cstMinusTwo = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getFloatAttr(resultElemTy, -2.0));
+    Value cstTwoPie = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getFloatAttr(resultElemTy, 2.0 * 3.14159));
+
+    // Construct the `linalg.generic` op.
+    auto resultRank = resultType.getRank();
+    SmallVector<AffineMap, 1> indexingMaps(
+        1, rewriter.getMultiDimIdentityMap(resultRank));
+    SmallVector<StringRef> iteratorTypes(resultRank,
+                                         getParallelIteratorTypeName());
+    Value randnRes =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, initTensor.getType(), /*inputs=*/ValueRange{},
+                /*outputs=*/initTensor, indexingMaps, iteratorTypes,
+                [&](OpBuilder &b, Location loc, ValueRange args) {
+                  Value temp = initialSeed;
+                  for (int i = 0; i < resultRank; i++) {
+                    Value index = b.create<linalg::IndexOp>(loc, i);
+                    temp = getNextTemp(b, loc, index, temp, multiplier,
+                                       incrementStep);
+                  }
+                  Value uOne = getUniformRandomNumber(b, loc, temp, min, max,
+                                                      resultElemTy);
+                  for (int i = 0; i < resultRank; i++) {
+                    Value index = b.create<linalg::IndexOp>(loc, i);
+                    temp = getNextTemp(b, loc, index, temp, multiplier,
+                                       incrementStep);
+                  }
+                  Value uTwo = getUniformRandomNumber(b, loc, temp, min, max,
+                                                      resultElemTy);
+                  // res = sqrt(-2ln(u1)) * cos(2*pie*u2)
```

Comment:
I think it would be simpler to do a decomposition. Something like this should work:

```python
def randn_decomp(size):
    u1 = torch.empty(size).uniform_()
    u2 = torch.empty(size).uniform_()
    return torch.sqrt(-2 * torch.log(u1)) * torch.cos(2 * 3.14159 * u2)
```

Reply:
That's not working. I have tried this, but it's giving the wrong result.

Reply:
This is the code that I added:
```
namespace {
class DecomposeAtenRandnOp : public OpRewritePattern<AtenRandnOp> {
public:
  using OpRewritePattern::OpRewritePattern;
  LogicalResult matchAndRewrite(AtenRandnOp op,
                                PatternRewriter &rewriter) const override {
    Location loc = op.getLoc();
    Type resultType = op.getType();

    Value none = rewriter.create<ConstantNoneOp>(loc);
    Value low = rewriter.create<Torch::ConstantFloatOp>(
        loc, rewriter.getF64FloatAttr((double)0.0));
    Value high = rewriter.create<Torch::ConstantFloatOp>(
        loc, rewriter.getF64FloatAttr((double)1.0));
    Value cstMinusTwo = rewriter.create<Torch::ConstantFloatOp>(
        loc, rewriter.getF64FloatAttr((double)-2.0));
    Value cstTwoPie = rewriter.create<Torch::ConstantFloatOp>(
        loc, rewriter.getF64FloatAttr((double)(2.0 * 3.14159)));

    Value emptyTensorA = rewriter.create<AtenEmptyMemoryFormatOp>(
        loc, resultType, op.size(), /*dtype=*/none, /*layout=*/op.layout(),
        /*device=*/op.device(), /*pin_memory=*/op.pin_memory(),
        /*memory_format=*/none);
    Value emptyTensorB = rewriter.create<AtenEmptyMemoryFormatOp>(
        loc, resultType, op.size(), /*dtype=*/none, /*layout=*/op.layout(),
        /*device=*/op.device(), /*pin_memory=*/op.pin_memory(),
        /*memory_format=*/none);

    Value uOne = rewriter.create<AtenUniformOp>(loc, resultType, emptyTensorA,
                                                /*from=*/low,
                                                /*to=*/high,
                                                /*generator=*/none);

    Value uTwo = rewriter.create<AtenUniformOp>(loc, resultType, emptyTensorB,
                                                /*from=*/low,
                                                /*to=*/high,
                                                /*generator=*/none);

    Value logUOne = rewriter.create<AtenLogOp>(loc, resultType, uOne);
    Value minusTwoLogUOne =
        rewriter.create<AtenMulScalarOp>(loc, resultType, logUOne, cstMinusTwo);
    Value r = rewriter.create<AtenSqrtOp>(loc, resultType, minusTwoLogUOne);
    Value theta =
        rewriter.create<AtenMulScalarOp>(loc, resultType, uTwo, cstTwoPie);
    Value cosTheta = rewriter.create<AtenCosOp>(loc, resultType, theta);
    rewriter.replaceOpWithNewOp<AtenMulTensorOp>(op, op.getType(), r, cosTheta);
    return success();
  }
};
} // namespace
```

Reply:
Yeah, this is strange. I've opened an issue to track this: https://github.com/llvm/torch-mlir/issues/1608